lib/Sema/CSApply.cpp

```diff
@@ -8148,6 +8149,95 @@ bool ConstraintSystem::applySolutionFix(
   }
+
+  case FixKind::AddConformace: {
+    auto getMissingConformance = [&](ConstraintLocator *locator) {
+      auto *anchor = locator->getAnchor();
+      auto &requirement = locator->getPath().back();
+      auto result = MissingConformances.find({anchor, requirement.getValue()});
+      assert(result != MissingConformances.end());
+      return result->getSecond();
+    };
+
+    auto conformance = getMissingConformance(locator);
+
+    auto *anchor = locator->getAnchor();
+    auto owner = solution.simplifyType(getType(anchor));
+
+    auto type = conformance.first;
+    auto protocolType = conformance.second->getDeclaredType();
+
+    //  Find `ApplyExpr` based on a function expression attached to it.
+    auto findApplyExpr = [](Expr *parent, Expr *fnExpr) -> ApplyExpr * {
+      ApplyExpr *applyExpr = nullptr;
+      parent->forEachChildExpr([&applyExpr, &fnExpr](Expr *subExpr) -> Expr * {
+        auto *AE = dyn_cast<ApplyExpr>(subExpr);
+        if (!AE || AE->getFn() != fnExpr)
+          return subExpr;
+
+        applyExpr = AE;
+        return nullptr;
+      });
+      return applyExpr;
+    };
+
+    auto getArgumentAt = [](ApplyExpr *AE, unsigned index) -> Expr * {
+      assert(AE);
+
+      auto *arg = AE->getArg();
+      if (auto *TE = dyn_cast<TupleExpr>(arg))
+        return TE->getElement(index);
+
+      assert(index == 0);
+      if (auto *PE = dyn_cast<ParenExpr>(arg))
+        return PE->getSubExpr();
+
+      return arg;
+    };
+
+    auto *applyExpr = findApplyExpr(expr, anchor);
+
+    Optional<unsigned> atParameterPos;
+    // Sometimes fix is recorded by type-checking sub-expression
+    // during normal diagnostics, in such case call expression
+    // is unavailable.
+    if (applyExpr) {
+      // If this is a static, initializer or operator call,
+      // let's not try to diagnose it here, but refer to expression
+      // diagnostics.
+      if (isa<BinaryExpr>(applyExpr) || isa<TypeExpr>(anchor))
+        return false;
+
+      if (auto *fnType = owner->getAs<AnyFunctionType>()) {
+        auto parameters = fnType->getParams();
+        for (auto index : indices(parameters)) {
+          if (parameters[index].getType()->isEqual(type)) {
+            atParameterPos = index;
+            break;
+          }
+        }
+      }
+    }
+
+    if (type->isExistentialType()) {
+      auto diagnostic = diag::protocol_does_not_conform_objc;
+      if (type->isObjCExistentialType())
+        diagnostic = diag::protocol_does_not_conform_static;
+
+      TC.diagnose(anchor->getLoc(), diagnostic, type, protocolType);
+    } else if (atParameterPos) {
+      // Requirement comes from one of the parameter types,
+      // let's try to point diagnostic to the argument expression.
+      auto *argExpr = getArgumentAt(applyExpr, *atParameterPos);
+      TC.diagnose(argExpr->getLoc(),
+                  diag::cannot_convert_argument_value_protocol, type,
+                  protocolType);
```

Comment:
Oh! Oh! In many cases, we can probably resolve`applyExpr->getFn()` down to a declaration, in which case we can produce a note pointing to the declaration we are trying to call. Bonus points for finding the type variable bindings when that declaration is generic. It would be super amazing if we could do something like this:

```
note: in call to function 'foo(blah:)' [with T=Int, U=String]
```

Reply:
Indeed! I've mentioned that in reply to Slava's comment - I'm going to change the diagnostic to `<decl-kind> <name> requires that 'X' conform to 'Y' [with T = X]`